---
title: "Assignment-3"
author: "Mandhri Abeysooriya and Rusheni Senanayake"
theme: cosmo
output:
  word_document: default
  pdf_document: default
  html_document: default
---


# PART-1.


## Question1-(Read in the file, making the gene accession numbers the row names. Show a table of values for the first six genes).

In order to address this question, initially download the given file "gene_expression.tsv‚Äù whichcontains RNA-seq count data for two samples of interest.
Use the command, download.file ("URL",destfile="name") to download "gene expression.tsv".destfile="name" command will be used to give the destination name for the downloaded file.
secondly,read the file using command, read.table("name of the file")
Then to make an object, the data set needed to be saved under a name in environment. 
In here the name of the object will be given as "x"
In oder to save the file as an object, run x<-read.table("geneexpression.tsv", header = TRUE) command.
Then look the first six genes using head(x) 
Full code use to solve the problem is,
 
```{r,chunck1,echo=TRUE}
download.file("https://raw.githubusercontent.com/markziemann/SLE712_files/master/bioinfo_asst3_part1_files/gene_expression.tsv",destfile = "geneexpression.tsv")

x<-read.table("geneexpression.tsv", header = TRUE)
head(x)
```

## Question2-(Make a new column which is the mean of the other columns. Show a table of values for the first sixgenes.)


An object named "meanofsamples" is saved in the environment using meanofthesample <- and then to find the means of each row, rowmeans() function is used. 
The reason for the use of rowMeans is that this fucntion will produce the mean in considering the every row where as columns are being selective.
rowmeans(x[,c(2:3)]) command will give an output where mean of the rows will be calculated by considering columns 2 and 3. 
Comma used as soon as square brackets open depicts the fact that this function consider all the rows.
Therfore full code can be written as x$meanofsamples<-rowMeans(x[,c(2:3)]) and to check the means for first six genes run x[1:6,]
Spefic columns on the table will be selected by using the syntax($) to run the command. 


Code to use in solving problem is,

```{r,chunck2,echo=TRUE}
meanofsamples<-rowMeans(x[,c(2:3)])
x$meanofsamples<-rowMeans(x[,c(2:3)])
x[1:6,]
```
 
## Question 3-(List the 10 genes with the highest mean expression)

Values in the column (meanofsamples) will be arranged in descending order and named as "ranks". decreasing argument will be TRUE.
This will be giving the results of the highest values to the lowest values of the column(meanofsamples) in reference to "GeneID"
Order() will results with ranking positions for each element.
Sort() will be sorting out the results in reference to "meanofsample" column. decreasing will be true to obtain highest value in numeric.
Since  the object "ranks" composed of interger data type, as,data frame(x[ranks,]) will be used to convert the integer data type to a data frame.
An object named "decended" will be created and first 10 rows of all the columns will be executed with the command, decended[1:10,]
This will be providing the highest mean values of the samples.
 
Following table show the first ten genes with highest genes.
 
```{r,chunck3,echo=TRUE}
ranks <-order(x$meanofsamples,decreasing = T)
ranks
sort(x$meanofsamples,decreasing = T,)
decended <-as.data.frame(x[ranks,])
decended[1:10,]
```
 

## Question 4-(Determine the number of genes with a mean <10)


To determine the nuber of genes which has mean value less than 10,subsetting has done. filtered() is selects or filters the rows of the data table that meet certain criteria creating a new data subset and subsetting was done using code filtered <- subset(x, meanofsamples > 10). Then to get the number used the function,nrow(filtered).nrow returns the number of rows that are present in a data frame or matrix. 

Codes use to solving problem is,

```{r,chunck4,echo=TRUE}
filtered <- subset(x, meanofsamples > 10)
nrow(filtered)
```

There were 15032 genes with means less than 10.

## Question 5 -(Make a histogram plot of the mean values in png format and paste it into your report.


Histogram were drawn using hist() function and for easy identification break it to 5 frequencies.histogram were drawn in mean of th samples in the x axis and frequancy was taken to y axis.To save the image in png format export the drawn graph and save the option under png format.

```{r,chunck5,echo=TRUE}
hist(x$meanofsample,main="distribution of mean",xlab="mean of samples",breaks=5)
```



## Question 6 -(Import this csv file into an R object. What are the column names?)


Initially run args() to find the arguments needed to read a csv file.
Then,In order to make an object, the data set needed to be saved under a name in environment.
So for that, the name of the object will be given as "R".
In oder to save the file as an object, run the command. R<-read.csv(file="tree_circumference,csv",header=true,sep=",").
Then run the command to depict the table in console by typing R .
To know the names of the columns in the cvs file, run the command, colnames(R).This step will provide with column names as characters.

```{r,chunck6,echo=TRUE}
args(read.csv)
R<-read.csv(file = "https://raw.githubusercontent.com/markziemann/SLE712_files/master/bioinfo_asst3_part1_files/growth_data.csv",header = TRUE,sep = ",")
R
colnames(R)
```


## Question 7-(Calculate the mean and standard deviation of tree circumference at the start and end of the study at both sites)


To calculate the mean and sd of the both sites, mean() and sd() are used. It shuould be noted that the column needed to be specified from the Table(R).
Then Make a subset for two sites(northeast and southwest). Name the subset as "Northeast" and "Southwest". 
In Northeast <-(R[R$Site=="northeast",]) - the name of the subset is Northeast. R is the table which include the data of cvs file.  Define the column of Interest in the table using the symbol($). ""syntax defines R$Site eqivalent to "northeast".
In Northeast <-(R[R$Site=="northeast",]) a comma after "northeast" suggest that the function is run in all columns.

mean and SD values of sites are,
```{r,chunck7,echo=TRUE}
#Northeast
Northeast <-(R[R$Site=="northeast",])
Northeast
mean(Northeast$Circumf_2004_cm)
sd(Northeast$Circumf_2004_cm)
mean(Northeast$Circumf_2019_cm)
sd(Northeast$Circumf_2019_cm)

#Southwest
Southwest <-(R[R$Site=="southwest",])
Southwest
mean(Southwest$Circumf_2004_cm)
sd(Southwest$Circumf_2004_cm)
mean(Southwest$Circumf_2019_cm)
sd(Southwest$Circumf_2019_cm)
```

## Question 8-(Make a box plot of tree circumference at the start and end of the study at both sites) 

Box plots were drawn for both the sites using  the function boxplot() to obtain the box plot for specific column.
Since it is for both the sites, it is better to run the function by selecting Northeast and Southwest objects instead R.


Box plots for northeast,

```{r,chunck8,echo=TRUE}
boxplot(Northeast$Circumf_2004_cm)
boxplot(Northeast$Circumf_2019_cm)
```

Boxplotsfor southwest,
```{r,chunck9,echo=TRUE}
boxplot(Southwest$Circumf_2004_cm)
boxplot(Southwest$Circumf_2019_cm)
```
### Making a one box plot 

Combining all the 4 box plots to a single box plot diagram. Use function xlab() to give the title for x-axis, ylab() for title to y-axis,main() is for title of the plot,Names() function is used to give names to each sectors.
boxplot(main="circumferences of trees in two sites",Northeast$Circumf_2004_cm,Northeast$Circumf_2019_cm,Southwest$Circumf_2004_cm,Southwest$Circumf_2019_cm,names=c("N2004","N2019","S2004","S2019"),ylab="Circumference(cm)",xlab="Trees")


```{r,chunck10,echo=TRUE}
boxplot(main="circumferences of trees in two sites",Northeast$Circumf_2004_cm,Northeast$Circumf_2019_cm,Southwest$Circumf_2004_cm,Southwest$Circumf_2019_cm,names=c("N2004","N2019","S2004","S2019"),ylab="Circumference(cm)",xlab="Trees")

```


## Question 9-(Calculate the mean growth over the past 10 years at each site.) 

### Creating the mean growth of the two subsets(Northeast and Southwest)

Run args() funtion to find the arguments needed for the rowMeans. The reason for the use of rowMeans is that this function will produce the mean in considering the every row where as columns will be selective.
Column 4 to column 6 are selected as they hold data for a 10 years time period.

```{r,chunck11,echo=TRUE}
args(rowMeans)
growth_Northeast <- rowMeans(Northeast[,c(4:6)])
growth_Northeast
growth_Southwest <- rowMeans(Southwest[,c(4:6)])
growth_Southwest
```

### Insert the new column to both the tables created under North_east and South_west

Run args() funtion to find the arguments needed for the rowMeans. Northeast$Circumf_2004
The reason for the use of rowMeans is that this fucntion will produce the mean in considering the every row where as selective columns are being sellected.
column 4 and column 6 are specified as they hold data for a 10 years time period.

```{r,chunck12,echo=TRUE}
Northeast$growth_Northeast <-rowMeans(Northeast[,c(4:6)])
Northeast
Southwest$growth_Southwest <-rowMeans(Southwest[,c(4:6)])
Southwest
```

## Question 10-(Use the t.test and wilcox.test functions to estimate the p-value that the 10 year growth is different at the two sites.)
Select the data frames which compose of mean of growth of two sites:growth_Northeast and growth_Southwest

### Application of t-test
The syntax used below is for a two sample t-test. The reason for use of two sample t-test is that samples of interest are collected from two sides.

```{r,chunck13,echo=TRUE}
t.test(growth_Northeast,growth_Southwest)
```

### Analysing t-test data
According to the p value which is 9.662e-06, it can be assume that the there is strong evidence aginst nulll hypothesis stating the fact that null hypothesis can  be rejected.
The mean of northeast is 23.29467 where as mena of southwest is 32.38867. Accoridngly, the mean of southwest is greater than northeast

### Application of Wilcox test
```{r,chunck14,echo=TRUE}
args(wilcox.test)
wilcox.test(growth_Northeast,growth_Southwest)
```
### Analysing wilcox data
```{r,chunck15,echo=TRUE}
wilcox.test(growth_Northeast,growth_Southwest)
```
A larer p value is obtained compared to t-test. Its assumed to be accurate as Wilcox.test relies onn different statstics.
Since Wilcox test is non parametric, the results tend to be less sensitive with more robust.




# Part 2

```{r,chunck16,echo=TRUE}
library("seqinr")
library("R.utils")
library("rBLAST")
library("ape")
library("ORFik")
library("Biostrings")
```

## Question 1-(Download the whole set of E. coli gene DNA sequences and use gunzip to decompress. Use the makeblast() function to create a blast database. How many sequences are present in the E.coli set?)


### Download the whole set of E. coli gene DNA sequences and use gunzip to decompress

Download the whole set of E. coli gene DNA sequences and use gunzip to decompress)
The link with fa.gv is selected as in the 1st question, it is mentioned to decompress and fa.gv extensioned file are compressed files.
So the link was selected and fa.gz extension was given to destile name.

```{r,chunck17,echo=TRUE}

download.file('ftp://ftp.ensemblgenomes.org/pub/bacteria/release-42/fasta/bacteria_0_collection/escherichia_coli_str_k_12_substr_mg1655/cds/Escherichia_coli_str_k_12_substr_mg1655.ASM584v2.cds.all.fa.gz',
              destfile = "ecoligene.fa.gz")
```

### uncompress the file
R.utils package was used whereas gunzip command assit in unzipping the document. The name of the file was given with the extension.
overwrite command was selected as True to keep the original file in the system.

```{r,chunck18,echo=TRUE}
R.utils::gunzip("ecoligene.fa.gz",overwrite=T)
```

### Making blast data base

```{r,chunck19,echo=TRUE}
makeblastdb("ecoligene.fa",dbtype = "nucl","-parse_seqids")
```


## Question 2-(Download the sample fasta sequences and read them in as above. For your allocated sequence, determine the length (in bp) and the proportion of GC bases).


### Download the sample fasta sequences and read them in as above. For the allocated sequence and determine the length (in bp) and the proportion of GC bases.
Download the sample fasta sequences and read them in as above. For your allocated sequence,
Determine the length (in bp) and the proportion of GC bases.
Downloading the file given in assigment pdf file which is of extension fa.
Since the file is of extension fa,it is clear that the file is already uncompressseed

```{r,chunck20,echo=TRUE}
download.file("https://raw.githubusercontent.com/markziemann/SLE712_files/master/bioinfo_asst3_part2_files/sample.fa",
              destfile = "sample.fa")
```

### read.fasta command is given to read the downloaded fasta sequences.

Downloaded fastas are saved in working directory as "samplefastas"
```{r,chunck21,echo=TRUE}
samplefastas <- read.fasta("sample.fa")
```

### one fasta sequence is been selected from the downloaded fasta sequence.
9th sequence is been selected and saved in the working directory as "Groupseq"

```{r,chunck22,echo=TRUE}
Groupseq <-samplefastas [[9]]
Groupseq
```

#length of 9th fasta seq
```{r,chunck23,echo=TRUE}
seqinr::getLength(Groupseq)
```

#length of GC propotion

```{r,chunck24,echo=TRUE}
seqinr::GC(Groupseq)
```

## Question 3-(You will be provided with R functions to create BLAST databases and perform blast searches. Use blast to identify what E. coli gene your sequence matches best. Show a table of the top 3 hits including percent identity, E-value and bit scores.You will be provided with R functions to create BLAST databases and perform blast searches. Use blast to identify what E. coli gene your sequence matches best. Show a table of the top 3 hits including percent identity, E-value and bit scores).


### making the necessary functions avaialble in environment

Make the sequences of commands avaialble in environment, which would allow the user to run blast

```{r,chunck25,echo=TRUE}
source("https://raw.githubusercontent.com/markziemann/SLE712_files/master/bioinfo_asst3_part2_files/mutblast_functions.R")
download.file("https://raw.githubusercontent.com/markziemann/SLE712_files/master/bioinfo_asst3_part2_files/mutblast_functions.R",destfile = "mublast.R")
```
select the myblastn_tab to run the blast.
run args() to identify the arguments needed in the command.
myseq=Groupseq whereas db="ecoligene.fa". 'db' will be based on which data base the sequence analysis needed to be run.
The results of the blast wil be saved as a vector named Groupseq_blast.

```{r,chunck26,echo=TRUE}
args(myblastn_tab)
Groupseq_blast <-myblastn_tab(myseq=Groupseq,db="ecoligene.fa")
str(Groupseq_blast)
head(Groupseq_blast)
```
### Details of the microbe

Name-heme export ABC transporter permease
Species-Escherichia coli str. K-12 substr. MG1655 
Location-Chromosome:2296362-2297024
Gene trees-EGGT00170000111466

## Qustion 4-(You will be provided with a function that enables you to make a set number of point mutations to your sequence of interest. Run the function and write an R code to check the number of mismatches between the original and mutated sequence.)


### Application of random mutations
Run the args() funtion to determine the arguments needed to be fulfilled in order to run the code.
Run length(Groupseq) to check the length of the sequence.
Sequence will be Groupseq in this question.
Complete the arguments for mutator function.
Save the mutations as a vector named Groupseq_mut. Note that, myseq will be the selected sequence of interest.
This can be of any number less than the length of the sequence.

```{r,chunck27,echo=TRUE}
length(Groupseq)
args(mutator)
Groupseq_mut<-mutator(myseq = Groupseq,nmut =100)
Groupseq_mut
```

### Creating a pairwise alignment
Groupseq_mut needed to be converted to a DNA string for the efficent storage and ease of manipulation of long data sequence.
c2s function will convert vector of chatacters to a single string.
PairwiseAlignment function is used to align the selected pairs of sequences.
In this case, it will be the sequences which have been converted to DNA strings.(Groupseq_mut,groupseq_).
pid function states the % simillarities/identities of two sequences.
alignment will be the object which stores the results of pairwiseAlignmnet of Groupseq_mut, groupseq_

```{r,chunck28,echo=TRUE}
Groupseq_mut <-DNAString(c2s(Groupseq_mut))
Groupseq_ <-DNAString(c2s(Groupseq))
alignment <-Biostrings::pairwiseAlignment(Groupseq_mut,Groupseq_)
pid(alignment)
alignment
```

### comparing the aligned sequences

Command 'comparestrings' will depicts the  altered positions between aligned sequences.
```{r,chunck29,echo=TRUE}
compareStrings(alignment)
```
### number of mismatch is calculated by nmismatch() fucntion.

```{r,chunck30echo=TRUE}
nmismatch(alignment)
```
According to the answer, it can be assumed that there are 75 mismatches in the mutated sequence(Groupseq_mut) in compared to original sequence of interest(Groupseq).


## Question5-(Using the provided functions for mutating and BLASTing a sequence, determine the number and proportion of sites that need to be altered to prevent the BLAST search from matching the gene of origin. Because the mutation is random, you may need to run this test multiple times to get a reliable answer.)

Using the provided functions for mutating and BLASTing a sequence, determine the number and
Proportion of sites that need to be altered to prevent the BLAST search from matching the gene of
Origin. Because the mutation is random, you may need to run this test multiple times to get a reliable answer.


## The basis of this question is to determinae the treshold of the blast so that the blast will be unable to recognise the gene from it's original sequence.

Running blast index
Creating a data base for the selected sequence 
Write.fasta function will be used to create a fasta sequece only for the sequence of interst.
In this question, Groupseq will be saved as a fa extension file and a data base of blast will be created based on the Groupseq.
dbtype is the nucleotides.

```{r,chunck31,echo=TRUE}
write.fasta(Groupseq,names="Groupseq",file.out = "Groupseq.fa")
makeblastdb(file="Groupseq.fa",dbtype = "nucl")
```

Groupseq-mut will be named as an object which will save the 10 ramdom mutations.  
data base of interest will be the data base created for Groupseq.
myblastn_tab will be the function to run the blast.Initially test with 10 random mutations and increase the mutations gradually.
Since the final answer needed to be the maximum number of mutations which will not recognise the mutated sequence with original sequence, the function is been applied. 
In this function, output will be given as 0 for null(mutated sequence is no longer recognised as a sequence related to originl sequence due to number of mutations  applied) and 1 when the mutated sequence is recognised as a mutated version of original sequence.
A new function named "my_blast_test is been declared where the signature of the function composed of myseq which is the sequrence of interest(mutated sequence),and nmut for number of randomm mutations needed to be applied to the sequence of interest.
The body of the function will be composing of the ----------- which will be needed to excute the function.

```{r,chunck32,echo=TRUE}
my_blast_test <- function(myseq,nmut) {
  
  Groupseq_mut <- mutator(myseq,nmut)
  
  test <- myblastn_tab(myseq = Groupseq_mut, db = "Groupseq.fa")
  
  if (is.null(test)) {
    result=0
  } else {
    result=1
  }
  
  return (result)
  
}
```

Application of the function(my_blast_test)
use the replicate to run the code many times.keep increasing nmuts
mean(res) will be giving the propotion of succesful blasts

```{r,chunck33,echo=TRUE}
res <- replicate(n=10, my_blast_test(myseq=Groupseq, nmut=10))
res
mean(res)

res <- replicate(n=10, my_blast_test(myseq=Groupseq, nmut=20))
res
mean(res)

res <- replicate(n=10, my_blast_test(myseq=Groupseq, nmut=30))
res
mean(res)

res <- replicate(n=10, my_blast_test(myseq=Groupseq, nmut=40))
res
mean(res)

res <- replicate(n=10, my_blast_test(myseq=Groupseq, nmut=50))
res
mean(res)
res <- replicate(n=10, my_blast_test(myseq=Groupseq, nmut=100))
res
mean(res)
res <- replicate(n=10, my_blast_test(myseq=Groupseq, nmut=150)) 
res
mean(res)
res <- replicate(n=10, my_blast_test(myseq=Groupseq, nmut=200)) 
res
mean(res)
res <- replicate(n=10, my_blast_test(myseq=Groupseq, nmut=250))
res
mean(res)
res <- replicate(n=10, my_blast_test(myseq=Groupseq, nmut=300))
res
mean(res)
```
#Number of sites that neede to be altered 250 bases.

## Question 6(Provide a chart or table that shows how the increasing proportion of mutated bases reduces the ability for BLAST to match the gene of origin. Summarise the results in 1 to 2 sentences).Provide a chart or table that shows how the increasing proportion of mutated bases reduces the ability for BLAST to match the gene of origin. Summarise the results in 1 to 2 sentences.

In this question,the propotion of succesul blast hits needed to be taken as a vector in order to draw a graph. 
In order to do, a fuction is declared naming run_trials which composed of myseq,NMUTS,Ntrials in the signature of the function.
In this function,the body will composed of the object "res" (n,replicate and my_blast_test).

```{r,chunck34,echo=TRUE}
run_trials<- function(myseq,NMUTS,Ntrials){
  replicate_response <-replicate(n=Ntrials,my_blast_test(myseq = myseq,nmut = NMUTS))
  res <-mean(replicate_response)
  return (res)
}
```


A vector will be declared which composed of number of random mutations which are needed to be applied to the gene of interest(Group_seq)
50,100,150,200,250,300 are used as number of  random mutations applied.

```{r,chunck35,echo=TRUE}
NMUTS<-c(50,100,150,200,250,300)
```

'for loop' is created to get the corresponding results related to 50,100,150,200,250,300
In order to do so, a vector was created which named mutation_list. Index was declared as 1 since the output values should not be replacing the previous values in muatation_list

An object named 'Runtrial_results was created which compose of run_trial function's results.

```{r,chunck36,echo=TRUE}
mutation_list<- c()
i=1
for (Numberof_mut in NMUTS) {
  print(Numberof_mut)
  Runtrial_results<-run_trials(myseq = Groupseq,NMUTS = Numberof_mut,Ntrials = 10)
  print(Runtrial_results)
  mutation_list[i]<-Runtrial_results
  i=i+1
}

mutation_list
```
### Plotting the graph for NMUTS=X and mutation_list=y
```{r,chunck37,echo=TRUE}
plot(NMUTS,mutation_list,xlab="Number of mutated sites",ylab="propotion of succcessful blasts",main="How elevation of mutated bases affect BLAST performance",type="l")

```

According to the graph, it can be concluded that the with increase of number of mutated bases in a sequence , the performance of the blast decreases gradually.Eventually,blast will no longer recognise the relationship between mutated sequence and the original sequence. In this graph,at 250 random mutations, the mutated sequence will be unidentical to the sequence of interest(original sequence).
